dnf/base.py

```diff
@@ -2520,6 +2518,25 @@ def _raise_package_not_found_error(self, pkg_spec, forms, reponame):
                 msg = _('Argument matched only with source RPMs')
             raise dnf.exceptions.PackageNotFoundError(msg, pkg_spec)
 
+    def _raise_package_not_installed_error(self, pkg_spec, forms, reponame):
+        all_query = self.sack.query(flags=hawkey.IGNORE_EXCLUDES).installed()
+        subject = dnf.subject.Subject(pkg_spec)
+        solution = subject.get_best_solution(
+            self.sack, forms=forms, with_src=False, query=all_query)
+
+        if not solution['query']:
+            raise dnf.exceptions.PackagesNotInstalledError(_('No match for argument'), pkg_spec)
+        if reponame is not None:
+            installed = [pkg for pkg in solution['query'] if self.history.repo(pkg) == reponame]
+        else:
+            installed = solution['query']
+        if not installed:
+            raise dnf.exceptions.PackagesNotInstalledError(
+                _('All matches were installed from a different repository for argument'), pkg_spec)
+        else:
+            msg = _('All matches were excluded by regular filtering for argument')
+            raise dnf.exceptions.PackagesNotInstalledError(msg, pkg_spec)
```

Comment:
Nitpick, you could put the message directly into the class constructor instead of the intermediate variable, the same way you do it in the `if` branch above it.

Reply:
Yes this would be an option, but it will change API. Exceptions are part of our API. DNF5

Reply:
No, you misunderstand I mean just do this:
```python
        else:
            raise dnf.exceptions.PackagesNotInstalledError(_('All matches were excluded by regular filtering for argument'), pkg_spec)
```

Again, the same way you do it in the `if` branch above it :) Anyway, just a nitpick too...